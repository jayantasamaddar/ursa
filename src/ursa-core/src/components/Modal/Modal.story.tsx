import React, { useState } from 'react';
import {
  ComponentStory,
  ComponentMeta,
  ComponentStoryFn
} from '@storybook/react';

import { Modal } from './Modal';
import { Button } from '../Button';

export default {
  title: 'Components/Modal',
  component: Modal
} as ComponentMeta<typeof Modal>;

const Template: ComponentStory<typeof Modal> = ({ children, ...args }) => {
  const [openModal, setOpenModal] = useState(false);

  const onSubmit = () => {
    setOpenModal(false);
    setTimeout(() => alert('Submitted'), 200);
  };

  return (
    <div className="Ursa-ModalProvider">
      <Button primary onClick={() => setOpenModal(true)}>
        Open
      </Button>

      <Modal
        isOpen={openModal}
        onClose={() => setOpenModal(false)}
        primaryButton={{ name: 'Submit', onClick: onSubmit }}
        {...args}
      >
        <div>
          <p>
            Use Instagram posts to share your products with millions of people.
            Let shoppers buy from your store without leaving Instagram.
          </p>
        </div>
      </Modal>
    </div>
  );
};

export const BasicModal: ComponentStoryFn<typeof Modal> = Template.bind({});

BasicModal.args = {
  title: 'Reach more shoppers with Instagram product tags'
};
