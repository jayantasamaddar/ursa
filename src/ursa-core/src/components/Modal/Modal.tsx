import React, { FC, ReactElement, MouseEvent, useEffect } from 'react';
import styled from '@emotion/styled';
import { ModalProps } from '../../types';

import { Portal } from '../Portal';
import { Spinner } from '../Spinner';
import { Backdrop } from '../Backdrop';
import { ModalHeader, ModalFooter } from './components';

const UrsaModal: FC<ModalProps> = ({
  title,
  className,
  children,
  primaryButton,
  secondaryButton = {},
  loading,
  isOpen,
  onClose
}): ReactElement | null => {
  /*********************************************************************************/
  // useEffect - Scrollbars
  /********************************************************************************/

  useEffect(() => {
    if (isOpen) {
      window.scrollTo(0, 0);
      if (document) {
        document.body.style.overflow = 'hidden';
        document.body.scroll = () => 'no';
      }
    }
    if (!isOpen && document) {
      document.body.style.overflow = 'auto';
      document.body.scroll = () => 'yes';
    }
  }, [isOpen]);

  /*********************************************************************************/
  // Render Modal JSX
  /********************************************************************************/
  if (!isOpen) return null;

  return (
    <Portal idPrefix="portal">
      <Backdrop />
      <div className={`Ursa-ModalContainer ${className || ''}`}>
        <div className="Ursa-Modal">
          <ModalHeader title={title} onClose={onClose} />

          {loading ? (
            <Spinner />
          ) : (
            <div className="Ursa-ModalContent">{children}</div>
          )}

          <ModalFooter
            primaryButton={primaryButton}
            secondaryButton={{ ...secondaryButton, onClick: onClose }}
          />
        </div>
      </div>
    </Portal>
  );
};

/*********************************************************************************/
// Render Styled Modal
/********************************************************************************/
export const Modal = styled(UrsaModal)(
  ({ theme: { color }, width, height, yPosition }) => `
  position: fixed;
  display: flex;
  justify-content: center;
  align-items: center;
  top: 0;
  left: 0;
  height: 100%;
  width: 100%;
  z-index: 9999;
  color: ${color['--ursa-text-primary']};
    
  .Ursa-Modal {
    display: flex;
    flex-direction: column;
    background-color: ${color['--ursa-bg-primary']};
    border-radius: 3px;
    width: ${width || '80%'};
    height: ${height || '80%'};
    padding: 20px;
    box-shadow: 0 20px 25px -5px rgb(30 41 59 / 0.4), 0 8px 10px -6px rgb(30 41 59 / 0.4);
    border-radius: 10px;
    overflow-y: auto;
    overscroll-behavior-y: contain;
    z-index: 9999;
  }
  .Ursa-ModalContent {
    display: flex;
    flex-direction: column;
    flex-grow: 1;
    justify-content: ${yPosition || 'flex-start'};
    padding: 20px;
    background-color: ${color['--ursa-bg-primary']};
  }
  `
);
