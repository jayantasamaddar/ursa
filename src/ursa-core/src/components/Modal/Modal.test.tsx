import React from 'react';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import { composeStories } from '@storybook/testing-react';
import * as stories from './Modal.story';

const { BasicModal } = composeStories(stories);

window.scrollTo = jest.fn();

describe('components/Modal', () => {
  describe('<BasicModal />', () => {
    afterAll(() => {
      jest.clearAllMocks();
    });
    it('Initial Render to the DOM', () => {
      render(<BasicModal />);
      const titleText = BasicModal.args?.title as string;

      /** Modal Open Button is in the Document */
      const modalOpenBtn = screen.getByText(/^Open$/); // Gets the Modal Open Button
      expect(modalOpenBtn).toBeInTheDocument(); // Is In the Document

      /** See if Modal Title is Present */
      const modalTitle = screen.queryByText(titleText);
      expect(modalTitle).toBeNull();
    });

    it('Run Snapshot Test', () => {
      const ModalEl = renderer.create(<BasicModal />).toJSON();
      expect(ModalEl).toMatchSnapshot();
    });

    it('Open Modal', () => {
      render(<BasicModal />);
      const titleText = BasicModal.args?.title as string;
      const modalOpenBtn = screen.getByText(/^Open$/); // Gets the Modal Open Button

      /** Open the Modal */
      fireEvent.click(modalOpenBtn);

      /** Modal Title is in the Document */
      const modalTitle = screen.getByText(titleText);
      const modalBodyText = screen.getByText(/^Use Instagram posts*/);
      const submitBtn = screen.getByRole('button', { name: /^Submit$/ }); // Submit Button
      const cancelBtn = screen.getByRole('button', { name: /^Cancel$/ }); // Cancel Button
      const closeBtn = screen.getByRole('button', { name: '' }); // Close Button
      const iconEl = screen.getByTestId('icon-external'); // Close Icon

      /** Modal Elements are in the Document */
      expect(modalTitle).toBeInTheDocument();
      expect(modalBodyText).toBeInTheDocument();
      expect(cancelBtn).toBeInTheDocument();
      expect(submitBtn).toBeInTheDocument();
      expect(closeBtn).toBeInTheDocument();
      expect(iconEl).toBeInTheDocument();
      expect(closeBtn).toContainElement(iconEl);
    });

    it('Cancel Button functionality', () => {
      render(<BasicModal />);
      const titleText = BasicModal.args?.title as string;
      const modalOpenBtn = screen.getByText(/^Open$/); // Gets the Modal Open Button

      /** Open the Modal */
      fireEvent.click(modalOpenBtn);

      const modalTitle = screen.getByText(titleText); // Get Modal Title
      const modalBodyText = screen.getByText(/^Use Instagram posts*/); // Body Text
      const submitBtn = screen.getByRole('button', { name: /^Submit$/ }); // Submit Button
      const cancelBtn = screen.getByRole('button', { name: /^Cancel$/ }); // Cancel Button
      const closeBtn = screen.getByRole('button', { name: '' }); // Close Button

      /** Click the Cancel Button */
      fireEvent.click(cancelBtn);

      /** Modal and Modal Elements should Disappear */
      expect(modalTitle).not.toBeInTheDocument();
      expect(modalBodyText).not.toBeInTheDocument();
      expect(submitBtn).not.toBeInTheDocument();
      expect(cancelBtn).not.toBeInTheDocument();
      expect(closeBtn).not.toBeInTheDocument();
      expect(modalOpenBtn).toBeInTheDocument();
    });

    it('Submit Button functionality', () => {
      render(<BasicModal />);
      const titleText = BasicModal.args?.title as string;
      const modalOpenBtn = screen.getByText(/^Open$/); // Gets the Modal Open Button

      /** Open the Modal */
      fireEvent.click(modalOpenBtn);

      const modalTitle = screen.getByText(titleText); // Get Modal Title
      const modalBodyText = screen.getByText(/^Use Instagram posts*/); // Body Text
      const submitBtn = screen.getByRole('button', { name: /^Submit$/ }); // Submit Button
      const cancelBtn = screen.getByRole('button', { name: /^Cancel$/ }); // Cancel Button
      const closeBtn = screen.getByRole('button', { name: '' }); // Close Button

      /** Click the Cancel Button */
      fireEvent.click(submitBtn);

      /** Modal and Modal Elements should Disappear */
      expect(modalTitle).not.toBeInTheDocument();
      expect(modalBodyText).not.toBeInTheDocument();
      expect(submitBtn).not.toBeInTheDocument();
      expect(cancelBtn).not.toBeInTheDocument();
      expect(closeBtn).not.toBeInTheDocument();
      expect(modalOpenBtn).toBeInTheDocument();
    });

    it('Close Button functionality', () => {
      render(<BasicModal />);
      const titleText = BasicModal.args?.title as string;
      const modalOpenBtn = screen.getByText(/^Open$/); // Gets the Modal Open Button

      /** Open the Modal */
      fireEvent.click(modalOpenBtn);

      const modalTitle = screen.getByText(titleText); // Get Modal Title
      const modalBodyText = screen.getByText(/^Use Instagram posts*/); // Body Text
      const submitBtn = screen.getByRole('button', { name: /^Submit$/ }); // Submit Button
      const cancelBtn = screen.getByRole('button', { name: /^Cancel$/ }); // Cancel Button
      const closeBtn = screen.getByRole('button', { name: '' }); // Close Button

      /** Click the Cancel Button */
      fireEvent.click(closeBtn);

      /** Modal and Modal Elements should Disappear */
      expect(modalTitle).not.toBeInTheDocument();
      expect(modalBodyText).not.toBeInTheDocument();
      expect(submitBtn).not.toBeInTheDocument();
      expect(cancelBtn).not.toBeInTheDocument();
      expect(closeBtn).not.toBeInTheDocument();
      expect(modalOpenBtn).toBeInTheDocument();
    });
  });
});
