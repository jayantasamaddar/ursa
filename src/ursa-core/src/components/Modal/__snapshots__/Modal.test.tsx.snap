// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Modal <BasicModal /> Run Snapshot Test 1`] = `
.emotion-0 {
  width: auto;
  min-width: 85px;
  padding-top: 0.875em;
  padding-bottom: 0.875em;
  padding-left: 1.5em;
  padding-right: 1.5em;
  font-size: 0.875rem;
  font-weight: bold;
  text-transform: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  border-width: 1px;
  border-style: solid;
  border-radius: 4px;
  border-color: rgba(20, 184, 166, 1);
  background-color: rgba(20, 184, 166, 1);
  color: white;
  transition-property: color,background-color,box-shadow,border-color;
  transition-duration: 0.2s;
  transition-timing-function: ease-in-out;
}

.emotion-0:hover {
  color: "auto";
  background-color: rgba(13, 148, 136, 1);
  border-color: rgba(13, 148, 136, 1);
  cursor: pointer;
  box-shadow: 0px 3px 1px -2px rgb(0 0 0 / 20%),0px 2px 2px 0px rgb(0 0 0 / 14%),0px 1px 5px 0px rgb(0 0 0 / 12%);
  white-space: nowrap;
}

.emotion-0 .Ursa-ButtonIconLabel {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  justify-content:center align-items: center;
  gap: 0.5em;
}

.emotion-0 .Ursa-Icon {
  margin: auto;
}

<div
  className=Ursa-ModalProvider
>
  <div
    className=Ursa-ButtonContainer
  >
    <button
      className=Ursa-Button emotion-0
      onClick={[Function]}
      role=button
      type=button
    >
      Open
    </button>
  </div>
</div>
`;
